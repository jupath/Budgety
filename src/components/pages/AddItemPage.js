import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startAddItem } from '../../actions/budget';
import AddItemForm from '../forms/AddItemForm';

export class AddItemPage extends Component {
  state = {
    itemTypeState: undefined,
  };

  componentWillMount() {
    this.checkUrlParams();
  }

  checkUrlParams = () => {
    const { itemType } = this.props;
    if (itemType === 'income') {
      this.setState({ itemTypeState: 'income' });
    } else if (itemType === 'expense') {
      this.setState({ itemTypeState: 'expense' });
    } else {
      this.props.history.push('/');
    }
  };

  handleAddItem = (item) => {
    this.props.addItem(item);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="container px-lg-7">
        <h2 className="pt-4">Add new {this.state.itemTypeState}</h2>
        <AddItemForm handleAddItem={this.handleAddItem} itemType={this.state.itemTypeState} />
      </div>
    );
  }
}

AddItemPage.propTypes = {
  addItem: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  itemType: PropTypes.string.isRequired,
};

const mapStateToProps = (state, props) => ({
  itemType: props.match.params.type,
});

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(startAddItem(item)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddItemPage);
