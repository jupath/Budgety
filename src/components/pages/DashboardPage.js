import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import BudgetSummary from '../layout/BudgetSummary';
import ListFilterForm from '../forms/ListFilterForm';
import ListHeader from '../list/ListHeader';
import BudgetList from '../list/BudgetList';
import itemsSelector from '../../selectors/items';

export const DashboardPage = props => (
  <div>
    <BudgetSummary />
    <ListFilterForm />
    <div className="container">
      <div className="row">
        <div className="col-sm-6">
          <ListHeader listName="Expenses" itemsType="expense" />
          {props.expenses.length > 0 && <BudgetList items={props.expenses} />}
        </div>
        <div className="col-sm-6">
          <ListHeader listName="Incomes" itemsType="income" />
          {props.incomes.length > 0 && <BudgetList items={props.incomes} />}
        </div>
      </div>
    </div>
  </div>
);

const mapStateToProps = state => ({
  incomes: itemsSelector(state.items, 'income', state.filters),
  expenses: itemsSelector(state.items, 'expense', state.filters),
});

DashboardPage.propTypes = {
  incomes: PropTypes.arrayOf(PropTypes.object),
  expenses: PropTypes.arrayOf(PropTypes.object),
};

DashboardPage.defaultProps = {
  incomes: [],
  expenses: [],
};

export default connect(mapStateToProps)(DashboardPage);
