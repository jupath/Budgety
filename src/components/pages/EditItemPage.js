import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import { startUpdateItem, startDeleteItem } from '../../actions/budget';
import AddItemForm from '../forms/AddItemForm';

export class EditItemPage extends Component {
  state = {
    modal: false,
  }

  onDeleteItem = () => {
    this.props.deleteItem(this.props.item.id);
    this.toggleModal();
    this.props.history.push('/');
  };

  toggleModal = () => {
    this.setState({
      modal: !this.state.modal,
    });
  }

  handleUpdateItem = (item) => {
    this.props.updateItem(this.props.item.id, item);
    this.props.history.push('/');
  }

  render() {
    return (
      <div className="container mt-4 px-lg-7">
        <h2>Update {this.props.item.itemType}</h2>
        <AddItemForm handleAddItem={this.handleUpdateItem} item={this.props.item} />
        <Button className="mt-2 px-4 py-2 text-uppercase" id="toggle-modal" color="danger" onClick={this.toggleModal}>Delete</Button>
        <Modal isOpen={this.state.modal} toggle={this.toggleModal}>
          <ModalHeader toggle={this.toggleModal}>Delete item</ModalHeader>
          <ModalBody>
            Are you sure you want to delete this?
          </ModalBody>
          <ModalFooter>
            <Button color="danger" id="delete" onClick={this.onDeleteItem}>Delete</Button>{' '}
            <Button color="secondary" id="cancel" onClick={this.toggleModal}>Cancel</Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

EditItemPage.propTypes = {
  updateItem: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
  item: PropTypes.objectOf(PropTypes.any).isRequired,
  history: PropTypes.object,
};

const mapStateToProps = (state, props) => ({
  item: state.items.find(item => item.id === props.match.params.id),
});

const mapDispatchToState = dispatch => ({
  updateItem: (id, item) => dispatch(startUpdateItem(id, item)),
  deleteItem: id => dispatch(startDeleteItem(id)),
});

export default connect(mapStateToProps, mapDispatchToState)(EditItemPage);
