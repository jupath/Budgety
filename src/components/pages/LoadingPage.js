import React from 'react';
import loader from '../../images/loading.svg';

const LoadingPage = () => (
  <div className="loader">
    <img src={loader} alt="loading..." />
  </div>
);

export default LoadingPage;
