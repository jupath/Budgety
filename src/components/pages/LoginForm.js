import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import { startSignInWithEmailAndPassword } from '../../actions/auth';

class LoginForm extends Component {
  state = {
    error: undefined,
  }

  doSignIn = (email, password) => {
    this.props.signInUser(email, password)
      .then(() => {
        if (this.props.authError) {
          this.setState({ error: this.props.authError });
        } else {
          this.setState({ error: undefined });
        }
      });
  }

  handleSignIn = (event) => {
    event.preventDefault();
    const email = event.target.elements.email.value;
    const password = event.target.elements.password.value;

    if (email === '' || password === '') {
      this.setState({ error: 'Please, fill out every field!' });
    } else {
      this.doSignIn(email, password);
    }
  }

  render() {
    return (
      <div>
        <Form onSubmit={this.handleSignIn}>
          <FormGroup>
            <Label className="sr-only" for="email">Email</Label>
            <Input
              type="email"
              name="email"
              id="email"
              placeholder="Email"
              required="required"
            />
          </FormGroup>
          <FormGroup>
            <Label className="sr-only" for="password">Password</Label>
            <Input
              type="password"
              name="password"
              placeholder="Password"
              required="required"
            />
          </FormGroup>
          { this.state.error && <p className="text-danger">{this.state.error}</p> }
          <Button color="success">Login</Button>
        </Form>
      </div>
    );
  }
}

LoginForm.propTypes = {
  signInUser: PropTypes.func.isRequired,
  authError: PropTypes.string,
};

LoginForm.defaultProps = {
  authError: undefined,
};

const mapStateToProps = state => ({
  authError: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  signInUser: (email, password) => dispatch(startSignInWithEmailAndPassword(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
