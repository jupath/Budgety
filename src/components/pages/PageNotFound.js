import React from 'react';

export const PageNotFound = () => (
  <div className="text-center pt-5">
    <h2>Ooops, error 404!</h2>
    <p>Page not found!</p>
  </div>
);

export default PageNotFound;
