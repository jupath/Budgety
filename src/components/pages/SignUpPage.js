import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import { startCreateUserWithEmailAndPass } from '../../actions/auth';
import { LOGIN } from '../../constants/routes';

class SignUpPage extends Component {
  state = {
    email: '',
    password1: '',
    password2: '',
    error: undefined,
  };

  onChangeEmail = (event) => {
    const email = event.target.value;
    this.setState({ email });
  }

  onChangePassword = (event) => {
    const password1 = event.target.value;
    this.setState({ password1 });
  }

  onChangePassword1 = (event) => {
    const password2 = event.target.value;
    this.setState({ password2 });
  }

  doCreateUser = (email, password) => {
    this.props.createUser(email, password)
      .then(() => {
        if (this.props.authError) {
          this.setState({ error: this.props.authError });
        } else {
          this.setState({ error: undefined });
        }
      });
  }

  handleCreateUser = (event) => {
    event.preventDefault();
    const { email, password1, password2 } = this.state;
    if (email === '' || password1 === '' || password2 === '') {
      this.setState({ error: 'Please, fill out every field!' });
    } else if (password1 !== password2) {
      this.setState({ error: 'The passwords are different!' });
    } else {
      this.doCreateUser(email, password1);
    }
  }

  render() {
    return (
      <div className="login">
        <div className="login__box mt-6">
          <div className="login__header pt-3 pb-2 px-4">
            <h3 className="text-center">Sign up</h3>
            <p className="mb-0">and start keeping track of your money</p>
          </div>
          <div className="login__body pt-4 px-3">
            <Form onSubmit={this.handleCreateUser}>
              <FormGroup>
                <Label className="sr-only" for="email">Email</Label>
                <Input
                  type="email"
                  name="email"
                  id="email"
                  placeholder="Email"
                  required="required"
                  onChange={this.onChangeEmail}
                />
              </FormGroup>
              <FormGroup>
                <Label className="sr-only" for="password1">Password 1</Label>
                <Input
                  type="password"
                  name="password1"
                  id="password1"
                  placeholder="Password"
                  required="required"
                  onChange={this.onChangePassword}
                />
              </FormGroup>
              <FormGroup>
                <Input
                  type="password"
                  name="password2"
                  id="password2"
                  placeholder="Password again"
                  required="required"
                  onChange={this.onChangePassword1}
                />
              </FormGroup>
              { this.state.error && <p className="text-danger">{this.state.error}</p> }
              <Button color="success">Sign up</Button>
            </Form>
            <p>Already have an account? <Link to={LOGIN}>Log in</Link></p>
          </div>
        </div>
      </div>
    );
  }
}

SignUpPage.propTypes = {
  createUser: PropTypes.func.isRequired,
  authError: PropTypes.string,
};

SignUpPage.defaultProps = {
  authError: undefined,
};

const mapStateToProps = state => ({
  authError: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  createUser: (email, password) => dispatch(startCreateUserWithEmailAndPass(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUpPage);
