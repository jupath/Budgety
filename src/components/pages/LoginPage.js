import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Button } from 'reactstrap';
import { startUserLogin } from '../../actions/auth';
import LoginForm from './LoginForm';
import { SIGNUP, RESET_PASS } from '../../constants/routes';

const LoginPage = ({ loginUser }) => (
  <div className="login">
    <div className="login__box mt-6">
      <div className="login__header pt-3 pb-2 px-4">
        <h3 className="text-center">Log in to your account</h3>
      </div>
      <div className="login__body pt-4 px-3">
        <Button className="login__google" onClick={loginUser}>
          Log in with Google
        </Button>
        <div className="login__separator-container py-3">
          <div className="login__separator" />
          <div className="px-2">OR</div>
          <div className="login__separator" />
        </div>
        <LoginForm />
        <p className="text-center">or <Link to={RESET_PASS}>Forgot password</Link></p>
        <p className="text-center">Don&rsquo;t have an account? <Link to={SIGNUP}>Sign up</Link></p>
      </div>
    </div>
  </div>
);

LoginPage.propTypes = {
  loginUser: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  loginUser: () => dispatch(startUserLogin()),
});

export default connect(null, mapDispatchToProps)(LoginPage);
