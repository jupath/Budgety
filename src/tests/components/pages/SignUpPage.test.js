import React from 'react';
import { shallow } from 'enzyme';
import { SignUpPage } from '../../../components/pages/SignUpPage';

let onSubitSpy, wrapper;

beforeEach(() => {
  onSubitSpy = jest.fn();
  wrapper = shallow(<SignUpPage createUser={onSubitSpy} />);
});

test('should render SingUpPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render SingUpPage correctly if there is an auth error', () => {
  const authError = 'Auth error';
  wrapper.setProps({ authError });
  expect(wrapper).toMatchSnapshot();
});

test('should set email in input change', () => {
  const value = 'test@example.com';
  wrapper.find('#email').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('email')).toBe(value);
});

test('should set password in input change', () => {
  const value = 'abc123';
  wrapper.find('#password1').simulate('change', {
    target: { value }
  });
  wrapper.find('#password2').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('password1')).toBe(value);
  expect(wrapper.state('password2')).toBe(value);
});

test('should handle create user submit', () => {
  const email = 'test@example';
  const password1 = 'abc123';
  const password2 = 'abc123';
  wrapper.setState({ email, password1, password2 });
  wrapper.find('Form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubitSpy).toHaveBeenLastCalledWith(email, password1);
});

test('should handle create user submit if the 2 passwords are different', () => {
  const email = 'test@example';
  const password1 = 'abc123';
  const password2 = '123abc';
  wrapper.setState({ email, password1, password2 });
  wrapper.find('Form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('The passwords are different!');
  expect(onSubitSpy).not.toHaveBeenCalled();
});
