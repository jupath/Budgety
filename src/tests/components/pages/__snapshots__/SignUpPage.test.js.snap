// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SingUpPage correctly 1`] = `
<div
  className="login"
>
  <div
    className="login__box mt-6"
  >
    <div
      className="login__header pt-3 pb-2 px-4"
    >
      <h3
        className="text-center"
      >
        Sign up
      </h3>
      <p
        className="mb-0"
      >
        and start keeping track of your money
      </p>
    </div>
    <div
      className="login__body pt-4 px-3"
    >
      <Form
        onSubmit={[Function]}
        tag="form"
      >
        <FormGroup
          tag="div"
        >
          <Label
            className="sr-only"
            for="email"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Email
          </Label>
          <Input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required="required"
            type="email"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            className="sr-only"
            for="password1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Password 1
          </Label>
          <Input
            id="password1"
            name="password1"
            onChange={[Function]}
            placeholder="Password"
            required="required"
            type="password"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Input
            id="password2"
            name="password2"
            onChange={[Function]}
            placeholder="Password again"
            required="required"
            type="password"
          />
        </FormGroup>
        <Button
          color="success"
          tag="button"
        >
          Sign up
        </Button>
      </Form>
      <p>
        Already have an account? 
        <Link
          replace={false}
          to="/"
        >
          Log in
        </Link>
      </p>
    </div>
  </div>
</div>
`;

exports[`should render SingUpPage correctly if there is an auth error 1`] = `
<div
  className="login"
>
  <div
    className="login__box mt-6"
  >
    <div
      className="login__header pt-3 pb-2 px-4"
    >
      <h3
        className="text-center"
      >
        Sign up
      </h3>
      <p
        className="mb-0"
      >
        and start keeping track of your money
      </p>
    </div>
    <div
      className="login__body pt-4 px-3"
    >
      <Form
        onSubmit={[Function]}
        tag="form"
      >
        <FormGroup
          tag="div"
        >
          <Label
            className="sr-only"
            for="email"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Email
          </Label>
          <Input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required="required"
            type="email"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            className="sr-only"
            for="password1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Password 1
          </Label>
          <Input
            id="password1"
            name="password1"
            onChange={[Function]}
            placeholder="Password"
            required="required"
            type="password"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Input
            id="password2"
            name="password2"
            onChange={[Function]}
            placeholder="Password again"
            required="required"
            type="password"
          />
        </FormGroup>
        <p
          className="text-danger"
        >
          Auth error
        </p>
        <Button
          color="success"
          tag="button"
        >
          Sign up
        </Button>
      </Form>
      <p>
        Already have an account? 
        <Link
          replace={false}
          to="/"
        >
          Log in
        </Link>
      </p>
    </div>
  </div>
</div>
`;
